docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+The authors propose the following list of possible reason codes for transitions. This list is not exhaustive, and can be further enhanced to capture any new transitions in the future.
+
+| Reason                                | Transition From State (SubState)                             | Transition To State (SubState)         |
+| ------------------------------------- | ------------------------------------------------------------ | -------------------------------------- |
+| `ClusterCreated`                      | nil                                                          | New                                    |
+| `DetectedPreviousCleanExit`           | New \| Started (Leader) \| Started (Follower)                | Initializing (DBValidationSanity)      |
+| `DetectedPreviousUncleanExit`         | New \| Started (Leader) \| Started (Follower)                | Initializing (DBValidationFull)        |
+| `DBValidationFailed`                  | Initializing (DBValidationSanity) \| Initializing (DBValidationFull) | Initializing (Restoration) \| New      |
+| `DBValidationSucceeded`               | Initializing (DBValidationSanity) \| Initializing (DBValidationFull) | Started (Leader) \| Started (Follower) |
+| `Initializing (Restoration)Succeeded` | Initializing (Restoration)                                   | Started (Leader)                       |
+| `ClusterScaledUp`                     | nil                                                          | New                                    |
+| `WaitingToJoinAsLearner`              | New                                                          | Starting (PendingLearner)              |
+| `JoinedAsLearner`                     | Starting (PendingLearner)                                    | Starting (Learner)                     |
+| `PromotedAsVotingMember`              | Starting (Learner)                                           | Started (Follower)                     |
+| `GainedClusterLeadership`             | Started (Follower)                                           | Started (Leader)                       |
+| `LostClusterLeadership`               | Started (Leader)                                             | Started (Follower)                     |
+
+#### API
+
+The authors propose to add the `EtcdMember` custom resource API to etcd-druid APIs and initially introduce it with `v1alpha1` version.
+
+```yaml
+apiVersion: druid.gardener.cloud/v1alpha1
+kind: EtcdMember
+metadata:
+  labels:
+    gardener.cloud/owned-by: <name of parent Etcd resource>
+  name: <name of the etcd-member>
+  namespace: <namespace | will be the same as that of parent Etcd resource>
+  ownerReferences:
+  - apiVersion: druid.gardener.cloud/v1alpha1
+    blockOwnerDeletion: true
+    controller: true
+    kind: Etcd
+    name: <name of the parent Etcd resource>
+    uid: <UID of the parent Etcd resource> 
+status:
+  id: <etcd-member id>
+  clusterID: <etcd cluster id>
+  peerTLSEnabled: <bool>
```

Comment:
Isn't `peerTLSEnabled` a good candidate for a `spec` too, e.g. having it in both places?

Reply:
We initially thought about it but then went against it. What we `define` as part of the spec is a desired state that must be constantly reconciled and eventually met and thereafter maintained. `peerTLSEnabled` is an information that is published by an `EtcdMember` depicting the as-is state of TLS enablement of peer URL as seen by the etcd process running within `etcd-wrapper` container within this etcd-member and thus belongs in status.

Reply:
I'm not sure if I got the point but if I read the DEP again then the conclusion would be that if TLS is enabled or not is derived from the `advertised peer URL(s)` and thus it doesn't make sense to have yet another spec field to configure `peerTLSEnabled`. Is that correct?

Reply:
Ok you are right, maybe we do not even need this any longer. Reason, recently we changed the handling of peer TLS URL changes and we do the following:
> _Assumption:_ Initial single-node etcd cluster does not have TLS enabled. When scale is requested then g/g enables TLS for peer communication

- We update the existing etcd-member and mount the TLS secret, this cases one restart of the pod.
- Once the update is over then we delete the pod, causing a second restart. This will now ensure that the existing pod's etcd process will deterministically always start with TLS enabled for peer URL.
- Now we scale etcd STS from 1-3.  At this point, 0th pod already has TLS enabled and therefore no other properly indicating and tracking TLS is required for other 2 pods that are going to be launched.

We can remove this property and also the section. Thanks for pointing it out.